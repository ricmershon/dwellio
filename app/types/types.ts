import { StructuredFormErrorMap } from "@/app/utils/build-form-error-map";

export interface Rates {
    nightly?: number;
    weekly?: number;
    monthly?: number
}

export type ActionState = {
    message?: string | null;
    status?: 'SUCCESS' | 'ERROR' | null,
    isFavorite?: boolean,
    isRead?: boolean,
    formData?: FormData,
    formErrorMap?: StructuredFormErrorMap
}

export interface PropertiesQuery {
    $or: [
        { name: RegExp },
        { description: RegExp },
        { amenities: RegExp },
        { type: RegExp },
        { 'location.street': RegExp },
        { 'location.city': RegExp },
        { 'location.state': RegExp },
        { 'location.zip': RegExp }
    ]
}

export interface PropertyImageData {
    secureUrl: string;
    publicId: string;
}

export type FormErrorsType = Record<string, string[]> | string[];

export const MAX_ITEMS_PER_PAGE = 6;