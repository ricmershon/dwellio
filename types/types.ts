import { StructuredFormErrorMap } from "@/utils/build-form-error-map";

export const ActionStatus = {
    INFO: 'info',
    SUCCESS: 'success',
    WARNING: 'warning',
    ERROR: 'error',
} as const;

type ActionStatus = (typeof ActionStatus)[keyof typeof ActionStatus];

export type ActionState = {
  message?: string | null;
  status?: ActionStatus | null;
  isFavorite?: boolean;
  isRead?: boolean;
  formData?: FormData;
  formErrorMap?: StructuredFormErrorMap;
};

export interface PropertiesQuery {
    $or: [
        { name: RegExp },
        { description: RegExp },
        { amenities: RegExp },
        { type: RegExp },
        { 'location.street': RegExp },
        { 'location.city': RegExp },
        { 'location.state': RegExp },
        { 'location.zip': RegExp }
    ]
}

export interface PropertyImageData {
    secureUrl: string;
    publicId: string;
    width: number;
    height: number;
}

export interface Rates {
    nightly?: number;
    weekly?: number;
    monthly?: number
}

export interface Amenity {
    id: string,
    value: string
}

export interface OptionType {
    value: string;
    label: string;
}

export type FormErrorsType = Record<string, string[]> | string[];

export const VIEWPORT_WIDTH_COOKIE_NAME = 'viewport_width';