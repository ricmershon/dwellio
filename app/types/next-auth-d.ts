// eslint-disable-next-line @typescript-eslint/no-unused-vars
import NextAuth, { DefaultSession } from "next-auth";

declare module 'next-auth' {
    interface Profile {
        picture?: string;
    }

    interface Session {
        user: {
            id?: string;
        } & DefaultSession["user"];
    }
}