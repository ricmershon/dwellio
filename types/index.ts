import { StructuredFormErrorMap } from "@/utils/build-form-error-map";

export const ActionStatus = {
    INFO: "info",
    SUCCESS: "success",
    WARNING: "warning",
    ERROR: "error",
} as const;

type ActionStatus = (typeof ActionStatus)[keyof typeof ActionStatus];

export type ActionState = {
    message?: string | null;
    status?: ActionStatus | null;
    isFavorite?: boolean;
    isRead?: boolean;
    formData?: FormData;
    formErrorMap?: StructuredFormErrorMap;
    userId?: string;
    isAccountLinked?: boolean;
    canSignInWith?: string[];
    email?: string;
    password?: string;
    shouldAutoLogin?: boolean;
};

export interface PropertiesQuery {
    $or: [
        { name: RegExp },
        { description: RegExp },
        { amenities: RegExp },
        { type: RegExp },
        { "location.street": RegExp },
        { "location.city": RegExp },
        { "location.state": RegExp },
        { "location.zip": RegExp }
    ]
}

export interface PropertyImageData {
    secureUrl: string;
    publicId: string;
    width: number;
    height: number;
}

export interface Rates {
    nightly?: number;
    weekly?: number;
    monthly?: number
}

export interface Amenity {
    id: string,
    value: string
}

export interface OptionType {
    value: string;
    label: string;
}

export type FormErrorsType = Record<string, string[]> | string[];

export interface AutocompleteFetchOptions {
    signal?: AbortSignal;
    includeDetails?: boolean;
    maxDetails?: number;
    regionCode?: string;
    includedPrimaryTypes?: string[];
};

export interface AutocompleteSuggestion {
    placePrediction: {
        placeId: string;
        text?: {
            text: string;
            languageCode?: string;
        };
        structuredFormat?: {
            mainText?: { text: string; matches?: Array<{ startOffset: number; endOffset: number }> };
            secondaryText?: { text: string; matches?: Array<{ startOffset: number; endOffset: number }> };
        };
    };
}

export interface AutocompletePrediction {
    placeId: string;
    text: string;
    structuredFormat?: AutocompleteSuggestion["placePrediction"]["structuredFormat"];
    street?: string | null;
    city?: string | null;
    state?: string | null;
    zipcode?: string | null;
}

export interface AutocompleteResponse {
    suggestions?: AutocompleteSuggestion[];
}

export interface AddressComponent {
    types?: string[];
    longText?: string;
    shortText?: string;
};

export interface PasswordValidation {
    isValid: boolean;
    errors: string[];
}

export const VIEWPORT_WIDTH_COOKIE_NAME = "viewport_width";