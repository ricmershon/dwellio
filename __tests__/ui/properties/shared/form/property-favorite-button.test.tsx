import React from 'react';
import { render, screen, fireEvent, waitFor } from '@/__tests__/test-utils';
import PropertyFavoriteButton from '@/ui/properties/shared/form/property-favorite-button';
import { ActionStatus } from '@/types/types';

// Mock dependencies
jest.mock('@/lib/actions/property-actions', () => ({
    getFavoriteStatus: jest.fn(),
    favoriteProperty: jest.fn(),
}));

// Create a complete mock that includes what layout.test.tsx needs plus what we need
const MockToastContainer = (props: { position?: string; theme?: string }) => (
    <div 
        data-testid="toast-container"
        data-position={props.position}
        data-theme={props.theme}
    />
);

jest.mock('react-toastify', () => ({
    ToastContainer: MockToastContainer,
    Slide: 'slide',
    toast: {
        error: jest.fn(),
        success: jest.fn(),
    },
}));

jest.mock('@heroicons/react/24/solid', () => ({
    HeartIcon: ({ className }: { className?: string }) => (
        <div data-testid="heart-solid-icon" className={className}>
            Solid Heart
        </div>
    ),
}));

jest.mock('@heroicons/react/24/outline', () => ({
    HeartIcon: ({ className }: { className?: string }) => (
        <div data-testid="heart-outline-icon" className={className}>
            Outline Heart
        </div>
    ),
}));

// Get mocked functions
const { getFavoriteStatus, favoriteProperty } = jest.mocked(
    jest.requireMock('@/lib/actions/property-actions')
);
const { toast } = jest.mocked(jest.requireMock('react-toastify'));

beforeEach(() => {
    jest.clearAllMocks();
    // Reset all mock implementations
    getFavoriteStatus.mockReset();
    favoriteProperty.mockReset();
    toast.error.mockReset();
    toast.success.mockReset();
});

describe('PropertyFavoriteButton', () => {
    const testPropertyId = 'test-property-123';

    describe('Initial Loading and State', () => {
        it('should render with outline heart initially', () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            expect(screen.getByTestId('heart-outline-icon')).toBeInTheDocument();
            expect(screen.queryByTestId('heart-solid-icon')).not.toBeInTheDocument();
        });

        it('should call getFavoriteStatus on mount', () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            expect(getFavoriteStatus).toHaveBeenCalledWith(testPropertyId);
            expect(getFavoriteStatus).toHaveBeenCalledTimes(1);
        });

        it('should display solid heart when property is favorited', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: true,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            await waitFor(() => {
                expect(screen.getByTestId('heart-solid-icon')).toBeInTheDocument();
            });

            expect(screen.queryByTestId('heart-outline-icon')).not.toBeInTheDocument();
        });

        it('should display outline heart when property is not favorited', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            await waitFor(() => {
                expect(screen.getByTestId('heart-outline-icon')).toBeInTheDocument();
            });

            expect(screen.queryByTestId('heart-solid-icon')).not.toBeInTheDocument();
        });
    });

    describe('Favorite Status Loading Error Handling', () => {
        it('should show toast error when getFavoriteStatus fails', async () => {
            const errorMessage = 'Failed to get favorite status';
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.ERROR,
                message: errorMessage,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            await waitFor(() => {
                expect(toast.error).toHaveBeenCalledWith(errorMessage);
            });
        });

        // TODO: Fix mock isolation issue - this test interferes with others
        it.skip('should throw error when getFavoriteStatus promise rejects', async () => {
            const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation();
            getFavoriteStatus.mockRejectedValue(new Error('Network error'));

            expect(() => {
                render(<PropertyFavoriteButton propertyId={testPropertyId} />);
            }).not.toThrow(); // Component should render, but useEffect error should be thrown

            consoleErrorSpy.mockRestore();
        });
    });

    describe('Form Structure', () => {
        it('should render form with correct classes', () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            const { container } = render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            const form = container.querySelector('form');
            expect(form).toBeInTheDocument();
            expect(form).toHaveClass('size-[18px]', 'z-20');
        });

        it('should render button with correct classes', () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            const button = screen.getByRole('button');
            expect(button).toBeInTheDocument();
            expect(button).toHaveClass('size-4', 'hover:cursor-pointer');
        });
    });

    describe('Heart Icon Styling', () => {
        it('should apply red color to solid heart icon', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: true,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            await waitFor(() => {
                expect(screen.getByTestId('heart-solid-icon')).toBeInTheDocument();
            });

            const solidHeart = screen.getByTestId('heart-solid-icon');
            expect(solidHeart).toHaveClass('text-red-600');
        });

        it('should apply gray color to outline heart icon', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            await waitFor(() => {
                const outlineHeart = screen.getByTestId('heart-outline-icon');
                expect(outlineHeart).toHaveClass('text-gray-600');
            });
        });
    });

    describe('Favorite Toggle Functionality', () => {
        it('should call favoriteProperty when button is clicked', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            favoriteProperty.mockResolvedValue({
                isFavorite: true,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            await waitFor(() => {
                expect(screen.getByTestId('heart-outline-icon')).toBeInTheDocument();
            });

            const button = screen.getByRole('button');
            fireEvent.click(button);

            expect(favoriteProperty).toHaveBeenCalledWith(testPropertyId);
        });

        it('should update heart icon after successful favorite toggle', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            favoriteProperty.mockResolvedValue({
                isFavorite: true,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            await waitFor(() => {
                expect(screen.getByTestId('heart-outline-icon')).toBeInTheDocument();
            });

            const button = screen.getByRole('button');
            fireEvent.click(button);

            await waitFor(() => {
                expect(screen.getByTestId('heart-solid-icon')).toBeInTheDocument();
            });

            expect(screen.queryByTestId('heart-outline-icon')).not.toBeInTheDocument();
        });

        it('should toggle from favorited to unfavorited', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: true,
            });

            favoriteProperty.mockResolvedValue({
                isFavorite: false,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            await waitFor(() => {
                expect(screen.getByTestId('heart-solid-icon')).toBeInTheDocument();
            });

            const button = screen.getByRole('button');
            fireEvent.click(button);

            await waitFor(() => {
                expect(screen.getByTestId('heart-outline-icon')).toBeInTheDocument();
            });

            expect(screen.queryByTestId('heart-solid-icon')).not.toBeInTheDocument();
        });
    });

    describe('Form Action Error Handling', () => {
        it('should handle favoriteProperty errors gracefully', async () => {
            const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation();

            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            favoriteProperty.mockRejectedValue(new Error('Favorite action failed'));

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            await waitFor(() => {
                expect(screen.getByTestId('heart-outline-icon')).toBeInTheDocument();
            });

            const button = screen.getByRole('button');
            
            // Should not crash when clicking
            expect(() => {
                fireEvent.click(button);
            }).not.toThrow();

            consoleErrorSpy.mockRestore();
        });
    });

    describe('Component Lifecycle', () => {
        it('should handle property ID changes', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            const { rerender } = render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            await waitFor(() => {
                expect(getFavoriteStatus).toHaveBeenCalledWith(testPropertyId);
            });

            const newPropertyId = 'new-property-456';
            rerender(<PropertyFavoriteButton propertyId={newPropertyId} />);

            await waitFor(() => {
                expect(getFavoriteStatus).toHaveBeenCalledWith(newPropertyId);
            });

            expect(getFavoriteStatus).toHaveBeenCalledTimes(2);
        });

        it('should cleanup properly on unmount', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            const { unmount } = render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            expect(() => {
                unmount();
            }).not.toThrow();
        });
    });

    describe('Accessibility', () => {
        it('should have accessible button element', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            const button = screen.getByRole('button');
            expect(button).toBeInTheDocument();
        });

        it('should be keyboard accessible', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            const button = screen.getByRole('button');
            button.focus();
            
            expect(button).toHaveFocus();
        });
    });

    describe('Snapshots', () => {
        it('should match snapshot when not favorited', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: false,
            });

            const { container } = render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            await waitFor(() => {
                expect(screen.getByTestId('heart-outline-icon')).toBeInTheDocument();
            });

            expect(container.firstChild).toMatchSnapshot();
        });

        it('should match snapshot when favorited', async () => {
            getFavoriteStatus.mockResolvedValue({
                status: ActionStatus.SUCCESS,
                isFavorite: true,
            });

            const { container } = render(<PropertyFavoriteButton propertyId={testPropertyId} />);

            await waitFor(() => {
                expect(screen.getByTestId('heart-solid-icon')).toBeInTheDocument();
            });

            expect(container.firstChild).toMatchSnapshot();
        });
    });
});