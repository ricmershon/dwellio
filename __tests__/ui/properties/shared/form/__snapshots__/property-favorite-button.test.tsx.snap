// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`PropertyFavoriteButton Snapshots should match snapshot when favorited 1`] = `
<form
  action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
  class="size-[18px] z-20"
>
  <button
    class="size-4 hover:cursor-pointer"
  >
    <div
      class="text-red-600"
      data-testid="heart-solid-icon"
    >
      Solid Heart
    </div>
  </button>
</form>
`;

exports[`PropertyFavoriteButton Snapshots should match snapshot when not favorited 1`] = `
<form
  action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
  class="size-[18px] z-20"
>
  <button
    class="size-4 hover:cursor-pointer"
  >
    <div
      class="text-gray-600"
      data-testid="heart-outline-icon"
    >
      Outline Heart
    </div>
  </button>
</form>
`;
